src/intents/Declaration.ts

```diff
@@ -101,6 +102,18 @@ class Declaration {
     return bits
   }
 
+  nonces (): BitNoncePair[] {
+    const bitNoncePairs: BitNoncePair[] = []
+    this.intents.forEach(intent => {
+      intent.nonces().forEach((pair) => {
+        if(!bitNoncePairs.find(existingPair => existingPair.nonce === pair.nonce)) {
```

Comment:
maybe unncessary optimitization but `find` is computationally expensive. for n intents, each with m nonce pairs, the worst-case time complexity is O(n * m^2). we could use `Set` instead which is O(1) for reading? 

this is a bit of a nitpick. it's only useful if we have to deal with a large number of nonces and intents so feel free to ignore 😇

Reply:
I think Set is not a good fit here, because has will use the object reference, and seems not trivial to compare values within the object

Reply:
i was thinking about sth like this:

```typescript
let seenNonces = new Set();

this.intents.forEach(intent => {
    intent.nonces().forEach((pair) => {
        if (!seenNonces.has(pair.nonce)) {
            seenNonces.add(pair.nonce);
            bitNoncePairs.push(pair);
        }
    });
});
```

are you saying that since pair.nonce could be located in another memory location, `has` won't check the uniqueness correctly?

Reply:
(btw this is not an important issue so don't let this block your PR, but happy to continue the discussion if u want)